modules/core/src/smithy4s/http/UrlForm.scala

```diff
@@ -19,123 +19,180 @@ package http
 
 import smithy4s.codecs.PayloadPath
 import smithy4s.codecs.PayloadPath.Segment
+import smithy4s.http.internals.UrlFormCursor
+import smithy4s.http.internals.UrlFormDataDecoder
+import smithy4s.http.internals.UrlFormDataDecoderSchemaVisitor
 import smithy4s.http.internals.UrlFormDataEncoder
 import smithy4s.http.internals.UrlFormDataEncoderSchemaVisitor
 import smithy4s.schema.CachedSchemaCompiler
 import smithy4s.schema.Schema
 
+import java.io.UnsupportedEncodingException
+import java.net.URLDecoder
 import java.net.URLEncoder
+import java.nio.CharBuffer
 import java.nio.charset.StandardCharsets
+import scala.collection.immutable.BitSet
 import scala.collection.mutable
 
-private[smithy4s] final case class UrlForm(
-    formData: UrlForm.FormData.MultipleValues
-) {
+private[smithy4s] final case class UrlForm(values: List[UrlForm.FormData]) {
+
   def render: String = {
     val builder = new mutable.StringBuilder
-    formData.writeTo(builder)
+    val lastIndex = values.size - 1
+    var i = 0
+    for (value <- values) {
+      value.writeTo(builder)
+      if (i < lastIndex) builder.append('&')
+      i += 1
+    }
     builder.result()
   }
 }
 
 private[smithy4s] object UrlForm {
 
-  sealed trait FormData extends Product with Serializable {
-    def prepend(segment: PayloadPath.Segment): FormData
-    def toPathedValues: Vector[FormData.PathedValue]
-    def widen: FormData = this
-    def writeTo(builder: mutable.StringBuilder): Unit
-  }
-  object FormData {
-    case object Empty extends FormData {
+  final case class FormData(path: PayloadPath, maybeValue: Option[String]) {
```

Comment:
Significant model simplification here. `FormData` is no longer a coproduct. The various places that need to handle it can do so with less code that's also more readable.